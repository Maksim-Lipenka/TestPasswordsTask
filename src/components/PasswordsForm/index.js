import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  SafeAreaView,
  ScrollView,
  TextInput,
  TouchableOpacity,
} from 'react-native';
import {connect} from 'react-redux';
import {compose} from 'redux';
import {Field, reduxForm, formValueSelector} from 'redux-form';
import {InputText, Button} from '../core';
import {atLeastOneCapitalLetter, atLeastOneNumber, atLeastOneRegularLetter, atLeastSixCharacters, required, } from '../../validation';
import { form, passwordConfirmationField, passwordConfirmationLabel, passwordField, passwordLabel, passwordRuleCapitalLetter, passwordRuleLength, passwordRuleMatches, passwordRuleNumber, passwordRuleRegularLetter, passwordRulesTitle, submitButtonLabel } from './strings';

const passwordsMatch = (value, allValues) =>
  value !== allValues.password ? "Passwords don't match" : undefined;

const PasswordsForm = props => {
  const {handleSubmit, passwordValue, passwordConfirmationValue} = props;

  const submit = values => {
    alert('Congratulations!');
  };

  return (
    <SafeAreaView style={styles.wrapper}>
      <View style={styles.container}>
        <Field
          name={passwordField}
          component={InputText}
          label={passwordLabel}
          validate={[required, atLeastOneRegularLetter, atLeastOneCapitalLetter, atLeastOneNumber, atLeastSixCharacters]}
          showIndicator
          secureTextEntry
        />
        <Field
          name={passwordConfirmationField}
          component={InputText}
          validate={[required, passwordsMatch]}
          label={passwordConfirmationLabel}
          showIndicator
          secureTextEntry
        />
        <View>
          <Text>{passwordRulesTitle}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastOneRegularLetter(passwordValue) && styles.indicatorValid]}>{passwordRuleRegularLetter}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastOneCapitalLetter(passwordValue) && styles.indicatorValid]}>{passwordRuleCapitalLetter}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastOneNumber(passwordValue) && styles.indicatorValid]}>{passwordRuleNumber}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastSixCharacters(passwordValue) && styles.indicatorValid]}>{passwordRuleLength}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !passwordsMatch(passwordConfirmationValue, { password: passwordValue }) && styles.indicatorValid]}>{passwordRuleMatches}</Text>
        </View>
        <Button
          onPress={handleSubmit(submit)}
          label={submitButtonLabel}
          disabled={!props.valid}
        />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    alignItems: 'center',
  },
  container: {
    height: '100%',
    width: 250,
    justifyContent: 'center',
    alignItems: 'center',
  },
  indicatorNotValid: {
    color: 'red'
  },
  indicatorValid: {
    color: 'green'
  }
});

export default compose(
  reduxForm({
    form: form,
  }),
  connect(state => ({
    passwordValue: formValueSelector(form)(state, passwordField),
    passwordConfirmationValue: formValueSelector(form)(state, passwordConfirmationField),
  })),
)(PasswordsForm);
