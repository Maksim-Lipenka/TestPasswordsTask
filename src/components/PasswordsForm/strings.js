export const form = 'passwords';

export const passwordField = 'password';
export const passwordConfirmationField = 'passwordConfirmation';

export const passwordLabel = 'Password';
export const passwordConfirmationLabel = "Password Confirmation"
export const submitButtonLabel = "Submit"

export const passwordRulesTitle = "Password rules:"
export const passwordRuleRegularLetter = "has at least one regular letter a-z"
export const passwordRuleCapitalLetter = "has at least one capital letter a-z"
export const passwordRuleNumber = "has at least one number 0-9"
export const passwordRuleLength = "length is greater than 6"
export const passwordRuleMatches = "Passwords matches"