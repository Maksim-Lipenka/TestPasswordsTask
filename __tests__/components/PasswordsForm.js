import 'react-native';
import React from 'react';
import PasswordsForm from '../../src/components/PasswordsForm';
import renderer from 'react-test-renderer';
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

const mockStore = configureMockStore();
const store = mockStore({});

describe('PasswordsForm', () => {
  const handleSubmit = jest.fn();

  const defaultProps = {handleSubmit: handleSubmit, passwordValue: "", passwordConfirmationValue: "", valid: false};

  it('renders correctly when not valid', () => {
    const tree = renderer.create(<Provider store={store}><PasswordsForm {...defaultProps} /></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when valid', () => {
    const tree = renderer.create(<Provider store={store}><PasswordsForm {...defaultProps} passwordValue="Password1" passwordConfirmationValue="Password1" /></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
