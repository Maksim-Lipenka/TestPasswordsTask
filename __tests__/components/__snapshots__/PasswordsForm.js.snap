// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordsForm renders correctly when not valid 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
        "width": 250,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text>
        Password
      </Text>
      <TextInput
        autoCapitalize="none"
        name="password"
        onBlur={[Function]}
        onChangeText={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "borderColor": "#1f1f1f",
              "borderWidth": 1,
              "height": 37,
              "marginBottom": 12,
              "paddingLeft": 8,
              "paddingRight": 32,
              "width": "100%",
            },
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 50,
              "bottom": 27.5,
              "height": 6,
              "position": "absolute",
              "right": 12,
              "width": 6,
            },
            Object {
              "backgroundColor": "green",
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text>
        Password Confirmation
      </Text>
      <TextInput
        autoCapitalize="none"
        name="passwordConfirmation"
        onBlur={[Function]}
        onChangeText={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "borderColor": "#1f1f1f",
              "borderWidth": 1,
              "height": 37,
              "marginBottom": 12,
              "paddingLeft": 8,
              "paddingRight": 32,
              "width": "100%",
            },
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 50,
              "bottom": 27.5,
              "height": 6,
              "position": "absolute",
              "right": 12,
              "width": 6,
            },
            Object {
              "backgroundColor": "green",
            },
          ]
        }
      />
    </View>
    <View>
      <Text>
        Password rules:
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        has at least one regular letter a-z
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        has at least one capital letter a-z
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        has at least one number 0-9
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        length is greater than 6
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        Passwords matches
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#3fd1a5",
          "height": 37,
          "justifyContent": "center",
          "marginTop": 10,
          "opacity": 1,
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        <Text>
          Submit
        </Text>
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`PasswordsForm renders correctly when valid 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
        "width": 250,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text>
        Password
      </Text>
      <TextInput
        autoCapitalize="none"
        name="password"
        onBlur={[Function]}
        onChangeText={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "borderColor": "#1f1f1f",
              "borderWidth": 1,
              "height": 37,
              "marginBottom": 12,
              "paddingLeft": 8,
              "paddingRight": 32,
              "width": "100%",
            },
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 50,
              "bottom": 27.5,
              "height": 6,
              "position": "absolute",
              "right": 12,
              "width": 6,
            },
            Object {
              "backgroundColor": "green",
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text>
        Password Confirmation
      </Text>
      <TextInput
        autoCapitalize="none"
        name="passwordConfirmation"
        onBlur={[Function]}
        onChangeText={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "borderColor": "#1f1f1f",
              "borderWidth": 1,
              "height": 37,
              "marginBottom": 12,
              "paddingLeft": 8,
              "paddingRight": 32,
              "width": "100%",
            },
            undefined,
          ]
        }
        value=""
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 50,
              "bottom": 27.5,
              "height": 6,
              "position": "absolute",
              "right": 12,
              "width": 6,
            },
            Object {
              "backgroundColor": "green",
            },
          ]
        }
      />
    </View>
    <View>
      <Text>
        Password rules:
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        has at least one regular letter a-z
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        has at least one capital letter a-z
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        has at least one number 0-9
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        length is greater than 6
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "red",
            },
            undefined,
          ]
        }
      >
        Passwords matches
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#3fd1a5",
          "height": 37,
          "justifyContent": "center",
          "marginTop": 10,
          "opacity": 1,
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        <Text>
          Submit
        </Text>
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
